src/software/ai/hl/stp/play/free_kick/free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,322 @@
+
+    if (num_defenders > 0)
+    {
+        defense_play->updateTactics(PlayUpdate(
+            event.common.world_ptr, num_defenders,
+            [&tactics_to_run](PriorityTacticVector new_tactics) {
+                for (const auto &tactic_vector : new_tactics)
+                {
+                    tactics_to_run.push_back(tactic_vector);
+                }
+            },
+            event.common.inter_play_communication,
+            event.common.set_inter_play_communication_fun));
+    }
+}
+
+void FreeKickPlayFSM::updateAlignToBallTactic(const WorldPtr &world_ptr)
+{
+    // Face towards the center of the field
+    Point position_to_face(0, 0);
+    Point ball_pos           = world_ptr->ball().position();
+    Vector direction_to_face = position_to_face - ball_pos;
+
+    align_to_ball_tactic->updateControlParams(
+        ball_pos - direction_to_face.normalize(ROBOT_MAX_RADIUS_METERS * 2),
+        direction_to_face.orientation(), 0);
+}
+
+bool FreeKickPlayFSM::shotFound(const Update &event)
+{
+    shot = calcBestShotOnGoal(event.common.world_ptr->field(),
+                              event.common.world_ptr->friendlyTeam(),
+                              event.common.world_ptr->enemyTeam(),
+                              event.common.world_ptr->ball().position(), TeamType::ENEMY);
+    return shot.has_value() &&
+           shot->getOpenAngle() >
+               Angle::fromDegrees(
+                   ai_config.attacker_tactic_config().min_open_angle_for_shot_deg());
+}
+
+void FreeKickPlayFSM::shootBall(const Update &event)
+{
+    LOG(INFO) << "Shooting ball.";
+    PriorityTacticVector tactics_to_run = {{}};
+
+    Point ball_pos = event.common.world_ptr->ball().position();
+
+    shoot_tactic->updateControlParams(
+        ball_pos, (shot->getPointToShootAt() - ball_pos).orientation(),
+        BALL_MAX_SPEED_METERS_PER_SECOND);
+    tactics_to_run[0].emplace_back(shoot_tactic);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+void FreeKickPlayFSM::startLookingForPass(const FreeKickPlayFSM::Update &event)
+{
+    pass_optimization_start_time = event.common.world_ptr->getMostRecentTimestamp();
+}
+
+bool FreeKickPlayFSM::timeExpired(const FreeKickPlayFSM::Update &event)
+{
+    Duration time_since_pass_optimization_start =
+        event.common.world_ptr->getMostRecentTimestamp() - pass_optimization_start_time;
+    return time_since_pass_optimization_start.toSeconds() >
+           ai_config.free_kick_play_config().max_time_commit_to_pass_seconds();
+}
+
+void FreeKickPlayFSM::chipBall(const Update &event)
+{
+    LOG(INFO) << "Time to look for pass expired. Chipping ball.";
+    PriorityTacticVector tactics_to_run = {{}};
+
+    Point ball_pos = event.common.world_ptr->ball().position();
+    std::optional<Robot> robot_kicking_opt =
+        event.common.world_ptr->friendlyTeam().getNearestRobot(ball_pos);
+    if (!robot_kicking_opt.has_value())
+    {
+        return;
```

Comment:
we need a LOG(WARNING) here because this may mean that we won't have a primitive for this tick which we ought to know about

Reply:
I added a warning here. If we do go into this if statement, it means that there are no friendly robot on the field, so not assigning any primitives would match.